ltrep.1
.\" keep in sync with ltrep.c
.\" see man-pages(7) for a style guide. also, always inspect the output of
.\" `groff -man -Tpdf ltrep.1` and make sure the right glyphs get produced,
.\" especially hyphens, dashes, quotes, apostrophes, tildes and carets.
.TH LTREP 1 2025-11-02 "LTREP 0.3" "User Commands"
.SH NAME
.\" don't use an em dash here
LTREP \- print lines matching a pattern
.SH SYNOPSIS
.\" bold is verbatim, italics is placeholder, roman is metacharacter
.B ltrep
.RI [ options ...]\&
.RB [ \-\- ]
.I pattern
.RI [ files ...]
.br
.B ltrep
.RI [ options ...]\&
.BR \-h , \-V
.SH DESCRIPTION
.PP
.B LTREP
searches
.I files
for lines matching
.IR pattern .
By default, a line is printed only when
.I pattern
matches the whole line.
.B LTREP
uses finite automata and guarantees linear-time searching.
.PP
A
.B \-\-
is needed when
.I pattern
begins in \(oq\-\(cq.
.PP
A file of \(oq\-\(cq denotes standard input. If no
.I files
are provided, read from standard input.
.PP
Show help and version info with
.BR \-h " and " \-V .
.SH OPTIONS
.PP
All options are unset by default. An option listed as
.RI \(oq \-a \(cq
is set by passing
.I \-a
and can\(cqt be unset, while an option listed as
.RI \(oq \-a / \-b \(cq
is set by passing
.I \-a
and is unset by passing
.IR \-b .
The flag supplied last takes precedence.
.TP
.B \-v
Invert match; print non-matching lines.
.TP
.BR \-p / \-x
Partial match; print lines that contain a match.
.br
Overrides
.BR \-o .
.TP
.B \-o
Match only; print the matches, not the lines.
.br
Matches may overlap so the output size may be quadratic in the input size.
.br
Overrides
.BR \-p / \-x .
.TP
.BR \-i / \-s
Ignore case; match case-insensitively.
.br
Overrides
.BR \-S .
.TP
.B \-S
Smart case; set
.B \-i
if pattern is all-lowercase.
.br
Overrides
.BR \-i / \-s .
.TP
.BR \-F / \-E
Parse the pattern as a fixed string, not a regex.
.TP
.BR \-H / \-h
Prefix matches with their file name.
.TP
.BR \-n / \-N
Prefix matches with their \%1-based line number.
.TP
.BR \-k / \-K
Prefix matches with their \%1-based column number.
.TP
.B \-b
Prefix matches with their \%0-based byte offset.
.TP
.BR \-T / \-t
Prefix matches with a \(rst so they line up.
.TP
.B \-c
Only print a count of matching lines per file.
.TP
.B \-l
Only print a list of files containing matches.
.TP
.B \-L
Only print a list of files containing no matches.
.TP
.B \-0
Terminate all file names with \(rs0, not : or \(rsn.
.TP
.BR \-z / \-Z
Use line terminator \(rs0 for input and output data.
.br
Overrides
.BR \-1 .
.TP
.B \-1
Use no line terminator for input and output data.
.br
Overrides
.BR \-z / \-Z .
.TP
.B \-q
Produce no output and prioritize exit status 0.
.\" keep in sync with patterns.md
.SH PATTERNS
.SS Notes
.PP
Matching is exact by default; for partial matching, surround a pattern with
.BR % s.
The metacharacter
.B %
is a shorthand for
.BR .* .
.PP
Whitespace is largely ignored; to match a whitespace character, use an escape
like
.BR \(rs\~ " or " \(rsn .
All metacharacters,
.BR \%\(rs\-.\(ti[]<>%{}*+?:|&=!(\~) ,
must be escaped to be matched literally, even inside character classes.
.PP
Character ranges can appear outside character classes. For example,
.B \%0\-9+
matches a non-empty sequence of digits. Character classes are complemented by
prefixing the opening bracket with
.BR \(ti .
For example,
.B \(ti[abc]
matches any character except
.BR a " or " b " or " c .
Single characters and character ranges can be complemented as well. For
example,
.B \(ti\(rsn*
matches a sequence of non-newline characters and
.B \%[aeiou\~\(tia\-z]
matches any character except lowercase consonants (insignificant whitespace
for clarity).
.PP
To match \(lqword\(rq characters (alphanumeric characters plus the
underscore),
you might use
.BR [\(rsm_] .
The metacharacter
.B .
matches any character, including newlines; to match any character except
newlines, you might use
.BR \(ti\(rsn .
To match patters like \(oqnumbers
.I separated by
commas\(cq you might use intercalation:
.B (\(rsd+)*!,
(redundant parentheses for clarity). To match patterns like \(oqidentifiers
.I which are not
keywords\(cq you might use intersection and
complementation:
.B \(rsa\(rsm*\~&\~!(if|do|for)
(insignificant whitespace for clarity).
.\" don't use title case for .SS
.SS Supported syntax
.TP
.I a
.I Literal Character
:: symbol
.br
Symbol consisting of the literal character
.IR a .
.br
Works for any printable character that\(cqs not a metacharacter.
.TP
.BI \(rs +
.I Metacharacter Escape
:: symbol
.br
Symbol consisting of the metacharacter
.IR \&+ .
.br
Works for any metacharacter,
.BR \%\(rs\-.\(ti[]<>%{}*+?:|&=!(\~) .
.TP
.BI \(rs r
.I Simple Escape
:: symbol
.br
Symbol associated with the simple escape
.B \(rs\c
.IR r .
.br
Available simple escapes are listed further below.
.TP
.BI \(rsx 41
.I Hexadecimal Escape
:: symbol
.br
Symbol with character code
.B 0x\c
.IR 41 .
.br
Takes exactly two digits.
.TP
any symbol
.I Symbol Promotion
:: symbol \-> symset
.br
Singleton set containing the symbol.
.TP
.IB a \- z
.I Character Range
:: (symbol, symbol) \-> symset
.br
Set of all characters from
.IR a " to " z
inclusively.
.br
Supports wraparound; for example,
.BR \%z\-a " means " \%\(tib\-y .
.TP
.B .
.I Symset Wildcard
:: symset
.br
Set of all characters.
.TP
.BI \(rs d
.I Shorthand
:: symset
.br
Set of characters in the shorthand
.B \(rs\c
.IR d .
.br
Available shorthands are listed further below.
.TP
.BI \(ti u
.I Symset Complement
:: symset \-> symset
.br
Set of all characters not in
.IR u .
.br
At most one may be applied per symset union/intersection
level.
.TP
.BI [ uv\c
.RB ... ]
.I Symset Union
:: [symset] \-> symset
.br
Set of characters in
.IR u " or in " v .
.TP
.BI < uv\c
.RB ... >
.I Symset Intersection
:: [symset] \-> symset
.br
Set of characters in
.IR u " and in " v .
.TP
any symset
.I Symset Promotion
:: symset \-> regex
.br
Length-one words over the symset.
.TP
.B %
.I Wildcard
:: regex
.br
All words.
.TP
.IB r {}
.I Empty Repetition
:: regex \-> regex
.br
Words some \%0-factoring of which consists of
.IR r s.
.TP
.IB r { 4 }
.I Exact Repetition
:: regex \-> regex
.br
Words some
.IR \%4 -\c
factoring of which consists of
.IR r s.
.TP
.IB r { 3 , 5 }
.I Bounded Repetition
:: regex \-> regex
.br
Words some
.IR \%3 \(en 5 -\c
factoring of which consists of
.IR r s.
.TP
.IB r { 3 ,}
.I Minimum Repetition
:: regex \-> regex
.br
Words some
.RI \%\(>= 3 -\c
factoring of which consists of
.IR r s.
.TP
.IB r {, 5 }
.I Maximum Repetition
:: regex \-> regex
.br
Words some
.RI \%\(<= 5 -\c
factoring of which consists of
.IR r s.
.TP
.IB r *
.I Kleene Star
:: regex \-> regex
.br
Words some \%n-factoring of which consists of
.IR r s.
.TP
.IB r +
.I Kleene Plus
:: regex \-> regex
.br
Words some \%\(>=1-factoring of which consists of
.IR r s.
.TP
.IB r ?
.I Optional
:: regex \-> regex
.br
Words some \%\(<=1-factoring of which consists of
.IR r s.
.TP
.IB r :{}
.I Dual Empty Repetition
:: regex \-> regex
.br
Words all \%0-factorings of which contain some
.IR r .
.TP
.IB r :{ 4 }
.I Dual Exact Repetition
:: regex \-> regex
.br
Words all
.IR \%4 -\c
factorings of which contain some
.IR r .
.TP
.IB r :{ 3 , 5 }
.I Dual Bounded Repetition
:: regex \-> regex
.br
Words all
.IR \%3 \(en 5 -\c
factorings of which contain some
.IR r .
.TP
.IB r :{ 3 ,}
.I Dual Minimum Repetition
:: regex \-> regex
.br
Words all
.RI \%\(>= 3 -\c
factorings of which contain some
.IR r .
.TP
.IB r :{, 5 }
.I Dual Maximum Repetition
:: regex \-> regex
.br
Words all
.RI \%\(<= 5 -\c
factorings of which contain some
.IR r .
.TP
.IB r :*
.I Dual Kleene Star
:: regex \-> regex
.br
Words all \%n-factorings of which contain some
.IR r .
.TP
.IB r :+
.I Dual Kleene Plus
:: regex \-> regex
.br
Words all \%\(>=1-factorings of which contain some
.IR r .
.TP
.IB r :?
.I Dual Optional
:: regex \-> regex
.br
Words all \%\(<=1-factorings of which contain some
.IR r .
.TP
.IB r* ! s
.I Intercalation
:: (regex, regex) \-> regex
.br
Words in
.I r*
but with
.I s
inserted between each
.IR r .
.br
Works for any quantifier,
.BR {}*+? .
.TP
.IB r : * ! s
.I Dual Intercalation
:: (regex, regex) \-> regex
.br
Words in
.IB r : *
but with
.I s
inserted between each
.IR r .
.br
Works for any quantifier,
.BR {}*+? .
.TP
.I rs
.I Concatenation
:: (regex, regex) \-> regex
.br
Words some \%2-factoring of which has head
.IR r " and tail " s .
.TP
.IB r : s
.I Dual Concatenation
:: (regex, regex) \-> regex
.br
Words all \%2-factorings of which have head
.IR r " or tail " s .
.TP
.IB r | s
.I Alternation
:: (regex, regex) \-> regex
.br
Words in
.IR r " or in " s .
.TP
.IB r & s
.I Intersection
:: (regex, regex) \-> regex
.br
Words in
.IR r " and in " s .
.TP
.IB r = s
.I Biconditional
:: (regex, regex) \-> regex
.br
Words in
.IR r " if and only if in " s .
.TP
.BI ! r
.I Complement
:: regex \-> regex
.br
Words not in
.IR r .
.br
At most one may be applied per grouping level.
.TP
.BI ( r )
.I Grouping
:: regex \-> regex
.br
Words in the subexpression
.IR r .
.PP
An \(oq\%n-factoring\(cq of a word is an \%n-tuple of strings whose
concatenation is that word. The empty word has a unique \%0-factoring,
namely the \%0-tuple; no other word has \%0-factorings.
.PP
Insignificant whitespace can appear anywhere except within escapes, character
ranges, shorthands, quantifiers, and intercalators.
.PP
Dual operators are dual with respect to complementation.
.SS Precedence
.PP
In decreasing order, and otherwise right associative:
.PP
.nr pd \n(PD
.nr PD 0
.IP "\h'1n'\(bu" 4n
Escapes;
.IP "\h'1n'\(bu" 4n
Character ranges, shorthands;
.IP "\h'1n'\(bu" 4n
Symset complementation;
.IP "\h'1n'\(bu" 4n
Quantification, dual quantification;
.IP "\h'1n'\(bu" 4n
Intercalation, dual intercalation;
.IP "\h'1n'\(bu" 4n
Concatenation;
.IP "\h'1n'\(bu" 4n
Dual concatenation;
.IP "\h'1n'\(bu" 4n
Complementation;
.IP "\h'1n'\(bu" 4n
Alternation, intersection, biconditional.
.nr PD \n(pd
.\" don't use title case for .SS
.SS Simple escapes
.TP
.B \(rsb
.I BACKSPACE
.br
.B \(rsx08
.TP
.B \(rsf
.I FORM FEED
.br
.B \(rsx0c
.TP
.B \(rsn
.I LINE FEED
.br
.B \(rsx0a
.TP
.B \(rsr
.I CARRIAGE RETURN
.br
.B \(rsx0d
.TP
.B \(rst
.I HORIZONTAL TABULATION
.br
.B \(rsx09
.TP
.B \(rsv
.I VERTICAL TABULATION
.br
.B \(rsx0b
.TP
.B \(rse
.I ESCAPE
.br
.B \(rsx1b
.PP
.SS Shorthands
.TP
.BR \(rsm ", " \(ti\(rsM
.I isalnum
(C89)
.br
.BR \%[0\-9A\-Za\-z] ,
.BR \%[\(rsd\(rsa] ,
.BR \%<\(rsg\(rsQ>
.TP
.BR \(rsa ", " \(ti\(rsA
.I isalpha
(C89)
.br
.BR \%[A\-Za\-z] ,
.BR \%[\(rsu\(rsl] ,
.BR \%<\(rsm\(rsD>
.TP
.BR \(rsk ", " \(ti\(rsK
.I isblank
(C99)
.br
.BR \%[\(rst\(rs\~] ,
.BR \%<\(rss\(ti\(rsn\-\(rsr>
.TP
.BR \(rsc ", " \(ti\(rsC
.I iscntrl
(C89)
.br
.BR \%[\(rsx00\-\(rsx1f\(rsx7f] ,
.BR \%<\(rsz\(rsP>
.TP
.BR \(rsd ", " \(ti\(rsD
.I isdigit
(C89)
.br
.BR \%0\-9 ,
.BR \%<\(rsm\(rsA>
.TP
.BR \(rsg ", " \(ti\(rsG
.I isgraph
(C89)
.br
.BR \%\(rs!\-\(rs\(ti ,
.BR \%[\(rsq\(rsm] ,
.BR \%<\(rsp\(ti\(rs\~>
.TP
.BR \(rsl ", " \(ti\(rsL
.I islower
(C89)
.br
.BR \%a\-z ,
.BR \%<\(rsa\(rsU>
.TP
.BR \(rsp ", " \(ti\(rsP
.I isprint
(C89)
.br
.BR \%\(rs\~\-\(rs\(ti ,
.BR \%[\(rsg\(rs\~] ,
.BR \%<\(rsz\(rsC>
.TP
.BR \(rsq ", " \(ti\(rsQ
.I ispunct
(C89)
.br
.BR \%[\(rs!\-/\(rs:\-@\(rs[\-\(ga\(rs{\-\(rs\(ti] ,
.BR \%<\(rsg\(rsM>
.TP
.BR \(rss ", " \(ti\(rsS
.I isspace
(C89)
.br
.BR \%[\(rst\-\(rsr\(rs\~] ,
.BR \%[\(rsk\(rsn\-\(rsr]
.TP
.BR \(rsu ", " \(ti\(rsU
.I isupper
(C89)
.br
.BR \%A\-Z ,
.BR \%<\(rsa\(rsL>
.TP
.BR \(rsh ", " \(ti\(rsH
.I isxdigit
(C89)
.br
.BR \%[0\-9A\-Fa\-f]
.TP
.BR \(rsz ", " \(ti\(rsZ
.I isascii
(BSD)
.br
.BR \%\(rsx00\-\(rsx7f ,
.BR \%[\(rsc\(rsp]
.SH "EXIT STATUS"
.PP
Exit status is:
.PP
.nr pd \n(PD
.nr PD 0
.TP
.B 2
if errors occurred; else,
.TP
.B 0
if a line matched; else,
.TP
.B 1
(no lines matched, no errors occurred).
.nr PD \n(pd
.PP
With
.B \-q
it is:
.PP
.nr pd \n(PD
.nr PD 0
.TP
.B 0
if a line matched; else,
.TP
.B 2
if errors occurred; else,
.TP
.B 1
(no lines matched, no errors occurred).
.nr PD \n(pd
.\" keep in sync with README.md
.SH EXAMPLES
.PP
Search for string literals in some C files, handling any escape sequences,
and print their line and column numbers:
.PP
.in +4n
.EX
.RB "$ " "ltrep \-Hnko \(aq\(dq(\(ti[\(rs\(rs\(dq]|\(rs\(rs.)*\(dq\(aq \c"
.B "ltrep.c ltre.c"
ltrep.c:1:10:\(dqltre.h\(dq
ltrep.c:19:14:\(dqv pxo isS FEHhnNkKb Ttc l L 0 zZq \(dq
ltrep.c:65:13:\(dqLTREP 0.3\(rsn\(dq
ltrep.c:66:14:\(dqLTREP --- print lines matching a pattern\(rsn\(dq
\&...
.EE
.in
.PP
Search for null-surrounded sequences of printable and whitespace characters,
somewhat like
.BR strings (1),
and print their offset into the file:
.PP
.in +4n
.EX
.RB "$ " "ltrep \-bz \(aq[\(rsp\(rss]{4,}\(aq bin/ltrep | \c"
.B "tr \(aq\(rs0\(rsn\(aq \(aq\(rsn\(rs0\(aq"
\&...
46832:v pxo isS FEHhnNkKb Ttc l L 0 zZq 
46872:unexpected nonprintable character
46906:LTREP 0.3
46917:parse error: %s near \(aq%.16s\(aq
46947:%s%c
\&...
.EE
.in
.
.\" vim:tw=78:ft=nroff:


ltrep.c
#include "ltre.h"
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef __unix__
#include <fcntl.h>
#include <sys/mman.h>
#include <unistd.h>
#endif

// steal implementation details. this probably has undefined behavior
struct dstate {
  struct dstate *transitions[256];
  bool accepting, terminating;
};

char *opts = "v pxo isS FEHhnNkKb Ttc l L 0 zZ1 q ";
struct args {
  struct {
    bool invert;  // -v
    bool partial; // -p/-x
    bool onlymat; // -o
    bool ignore;  // -i/-s
    bool smart;   // -S
    bool fixed;   // -F
    bool filehd;  // -H/-h
    bool lineno;  // -n/-N
    bool colno;   // -k/-K
    bool byteoff; // -b
    bool inittab; // -T/-t
    bool count;   // -c
    bool list;    // -l
    bool nlist;   // -L
    bool nulname; // -0
    bool nulline; // -z/-Z
    bool oneline; // -1
    bool quiet;   // -q
  } opts;
  char *pattern; // <pattern>
  char **files;  // [files...]
};

enum { EXIT_MATCH, EXIT_NOMATCH, EXIT_ERROR };

// the fundamental problem with -1 is quoting: in the output stream, how do you
// communicate that one match ends and the next one begins? let's take a closer
// look at line terminators (as in, \n when -Z and \0 when -z). with -x, line
// terminators terminate lines, each line is a word (a formal-language-theory
// word), and we output matching words. -p is the same but the pattern is first
// surrounded by a pair of wildcards. but for -o, there are two equivalent
// perspectives:
//  1. the input stream is a sequence of lines, line terminators terminate each
//     such line, each line is a string, and we output the substrings that are
//     matching words.
//  2. the input stream is a homogeneous blob of data, line terminators are
//     sentinel bytes, and the user asks that any match containig such sentinels
//     be ignored so we can use them to delineate matches in the output stream.
// perspective 2 gives us a -1 with quoting. matches that contain \0 bytes (\n
// bytes) are ignored, but it must be so because \0 bytes (\n bytes) are what's
// used in the output to delineate matches. for quoting matches that contain
// both \n and \0 bytes, you might use tr(1).

// keep in sync with ltrep.1
#define VER "LTREP 0.3\n"
#define INV "Unrecognized option '-%.*s'.\n"
#define TRY "Try 'ltrep -h' for more information.\n"
#define NAME "LTREP --- print lines matching a pattern\n"
#define USAGE                                                                  \
  "Usage:\n"                                                                   \
  "  ltrep [options...] [--] <pattern> [files...]\n"                           \
  "  ltrep [options...] -h,-V\n"
#define OPTS                                                                   \
  "Options:\n"                                                                 \
  "  -v     Invert match; print non-matching lines.\n"                         \
  "  -p/-x  Partial match; print lines that contain a match.\n"                \
  "  -o     Match only; print the matches, not the lines.\n"                   \
  "  -i/-s  Ignore case; match case-insensitively.\n"                          \
  "  -S     Smart case; set '-i' if pattern is all-lowercase.\n"               \
  "  -F/-E  Parse the pattern as a fixed string, not a regex.\n"               \
  "  -H/-h  Prefix matches with their file name.\n"                            \
  "  -n/-N  Prefix matches with their 1-based line number.\n"                  \
  "  -k/-K  Prefix matches with their 1-based column number.\n"                \
  "  -b     Prefix matches with their 0-based byte offset.\n"                  \
  "  -T/-t  Prefix matches with a \\t so they line up.\n"                      \
  "  -c     Only print a count of matching lines per file.\n"                  \
  "  -l     Only print a list of files containing matches.\n"                  \
  "  -L     Only print a list of files containing no matches.\n"               \
  "  -0     Terminate all file names with \\0, not : or \\n.\n"                \
  "  -z/-Z  Use line terminator \\0 for input and output data.\n"              \
  "  -1     Use no line terminator for input and output data.\n"               \
  "  -q     Produce no output and prioritize exit status 0.\n"
#define STATUS                                                                 \
  "Exit status is 2 if errors occurred, else 0 if\n"                           \
  "a line matched, else 1; with '-q' it is 0 if a\n"                           \
  "line matched, else 2 if errors occurred, else 1.\n"
#define HELP                                                                   \
  "" NAME "\n"                                                                 \
  "LTREP searches files for lines matching a pattern.\n"                       \
  "By default, a line is printed only when the pattern\n"                      \
  "matches the whole line. LTREP uses finite automata\n"                       \
  "and guarantees linear-time searching.\n"                                    \
  "\n" USAGE "\n"                                                              \
  "A '--' is needed when the pattern begins in '-'.\n"                         \
  "A file of '-' denotes standard input. If no files\n"                        \
  "are provided, read from standard input.\n"                                  \
  "Show help and version info with '-h' and '-V'.\n"                           \
  "\n" OPTS "\n"                                                               \
  "Options '-i/-s' and '-S' override eachother.\n"                             \
  "Options '-p/-x' and '-o' override eachother.\n"                             \
  "Options '-z/-Z' and '-1' override eachother.\n"                             \
  "Matches may overlap so when '-o' is supplied the\n"                         \
  "output size may be quadratic in the input size.\n"                          \
  "\n" STATUS ""

struct args parse_args(char **argv) {
  struct args args = {0};

  if (!*++argv)
    fputs(NAME TRY, stderr), exit(EXIT_ERROR);

  for (; *argv && **argv == '-'; argv++) {
    if (strcmp(*argv, "--") == 0 && argv++)
      break;
    if (strcmp(*argv, "-h") == 0 && !argv[1])
      fputs(HELP, stdout), exit(EXIT_SUCCESS);
    if (strcmp(*argv, "-V") == 0 && !argv[1])
      fputs(VER, stdout), exit(EXIT_SUCCESS);

    for (char *p, *opt = *argv + 1; *opt; opt++) {
      if ((p = strchr(opts, *opt)) && *opt != ' ')
        ((bool *)&args.opts)[p - opts >> 1] = !(p - opts & 1);
      else
        fprintf(stderr, INV TRY, *opt == '-' ? -1 : 1, opt), exit(EXIT_ERROR);

      args.opts.smart &= *opt != 'i' && *opt != 's';   // '-i/-s' override '-S'
      args.opts.onlymat &= *opt != 'p' && *opt != 'x'; // '-p/-x' override '-o'
      args.opts.oneline &= *opt != 'z' && *opt != 'Z'; // '-z/-Z' override '-1'
    }
  }

  if (!*argv)
    fputs(USAGE TRY, stderr), exit(EXIT_ERROR);
  args.pattern = *argv++;
  static char *read_stdin[] = {"-", NULL};
  args.files = *argv ? argv : read_stdin;

  args.opts.partial |= args.opts.onlymat; // '-o' overrides '-p/-x'
  args.opts.ignore |= args.opts.smart;    // '-S' overrides '-i/-s'
  args.opts.nulline |= args.opts.oneline; // '-1' overrides '-z/-Z'

  // not trying to be clever here. /\D/ and /\x6A/, for example, are treated as
  // uppercase and cause matches to become case sensitive. probably not much of
  // an issue because one could write /~\d/ and /\x6a/ instead
  if (args.opts.smart)
    for (char *c = args.pattern; *c; c++)
      args.opts.ignore &= !isupper(*c);

  return args;
}

int main(int argc, char **argv) {
  struct args args = parse_args(argv);

  char *error = NULL, *loc = args.pattern;
  struct regex *regex =
      args.opts.fixed ? ltre_fixed_string(loc) : ltre_parse(&loc, &error);
  if (error)
    fprintf(stderr, "parse error: %s near '%.16s'\n", error, loc),
        exit(EXIT_ERROR);

  // swapping checks for `args.partial` and `args.ignore` would not affect the
  // accepted language, but swapping checks for `args.partial` and `args.invert`
  // or swapping checks for `args.ignore` and `args.invert` would. we check
  // for `args.invert` last to preserve that:
  //   - `ltrep -vp` means 'does not contain'
  //   - `ltrep -vi` means 'is not a case variation of'
  //   - `ltrep -vpi` means 'does not contain any case variation of'
  //
  // given a regex /abc/, for match boundary extraction (when '-o' is supplied
  // and matching is not inverted), we construct:
  //  1. the regular partial DFA `dfa` that matches /%abc%/, which we run in
  //     the usual way to filter out lines that don't contain matches;
  //  2. a "reverse DFA" `rev_dfa` that matches /%cba/, which we run from the
  //     end of a line to find the spots where a partial match can begin;
  //  3. a "forward DFA" `fwd_dfa` that matches /abc/, which we run from the
  //     beginning of a partial match to find the spots where it can end.
  // so we perform one forward scan, then one backward scan within which we
  // perform several forward scans. the alternative would be to construct:
  //  1. a DFA that matches /%abc/, which we run forward from the start of a
  //     line to find the spots where a partial match can end; and
  //  2. a DFA that matches /cba/, which we run backward from the end of a
  //     partial match to find the spots where it can begin.
  // but that means performing one forward scan within which we perform several
  // backward scans, and that's not as nice to the prefetcher

  struct dstate *rev_dfa = NULL, *fwd_dfa = NULL;

  if (args.opts.ignore)
    regex = regex_ignorecase(regex, false);
  if (args.opts.onlymat && !args.opts.invert && !args.opts.count &&
      !args.opts.list && !args.opts.nlist) {
    fwd_dfa = ltre_compile(regex_incref(regex));
    rev_dfa = ltre_compile(regex_reverse(
        regex_concat(REGEXES(regex_incref(regex), regex_univ()))));
  }
  if (args.opts.partial)
    regex = regex_concat(REGEXES(regex_univ(), regex, regex_univ()));
  if (args.opts.invert)
    regex = regex_compl(regex);

  struct dstate *dfa = ltre_compile(regex);

  // be extremely careful with -o: in general, the space and time complexity
  // becomes quadratic in the input length. to preserve linear-time, linear-
  // space guarantees it is sufficient (and necessary, when other options don't
  // suppress any output) to show, for a given pattern, that there is a K such
  // that every input character is contained in at most K distinct matches. for
  // patterns with bounded match length this is always the case.

#define OUTPUT_MATCH /* args.opts, file, lineno, lineoff, line, begin, end */  \
  do {                                                                         \
    if (args.opts.count || args.opts.list || args.opts.nlist)                  \
      break;                                                                   \
    if (args.opts.quiet)                                                       \
      break;                                                                   \
                                                                               \
    if (args.opts.filehd)                                                      \
      printf("%s%c", *file, args.opts.nulname ? '\0' : ':');                   \
    if (args.opts.lineno)                                                      \
      printf("%zu:", lineno + 1);                                              \
    if (args.opts.colno)                                                       \
      printf("%zu:", begin - line + 1);                                        \
    if (args.opts.byteoff)                                                     \
      printf("%zu:", begin - line + lineoff);                                  \
    if (args.opts.inittab)                                                     \
      putchar('\t');                                                           \
    fwrite(begin, sizeof(*begin), end - begin, stdout);                        \
    args.opts.oneline ? 0 : putchar(args.opts.nulline ? '\0' : '\n');          \
  } while (0)

#define OUTPUT_LINE /* args.opts, file, lineno, lineoff, line, len, dfa */     \
  do {                                                                         \
    if (args.opts.count || args.opts.list || args.opts.nlist)                  \
      break;                                                                   \
    if (args.opts.quiet)                                                       \
      break;                                                                   \
                                                                               \
    if (args.opts.onlymat && !args.opts.invert) {                              \
      uint8_t *begin = line + len; /* rightmost to leftmost */                 \
      for (struct dstate *dstate = rev_dfa;;                                   \
           dstate = dstate->transitions[*--begin]) {                           \
        if (dstate->accepting) {                                               \
          uint8_t *end = begin; /* shortest to longest */                      \
          for (struct dstate *dstate = fwd_dfa;;                               \
               dstate = dstate->transitions[*end++]) {                         \
            if (dstate->accepting)                                             \
              OUTPUT_MATCH;                                                    \
            else if (dstate->terminating)                                      \
              break;                                                           \
            if (end == line + len)                                             \
              break;                                                           \
          }                                                                    \
        } else if (dstate->terminating)                                        \
          break;                                                               \
        if (begin == line)                                                     \
          break;                                                               \
      }                                                                        \
    } else {                                                                   \
      uint8_t *begin = line, *end = line + len;                                \
      OUTPUT_MATCH;                                                            \
    }                                                                          \
  } while (0)

#define OUTPUT_FILE /* args.opts, file, lineno, lineoff, count */              \
  do {                                                                         \
    if (!args.opts.count && (count ? !args.opts.list : !args.opts.nlist))      \
      break;                                                                   \
    if (args.opts.quiet)                                                       \
      break;                                                                   \
                                                                               \
    if (args.opts.filehd)                                                      \
      printf("%s%c", *file, args.opts.nulname ? '\0' : ':');                   \
    if (args.opts.lineno)                                                      \
      printf("%zu:", lineno + 1);                                              \
    if (args.opts.colno)                                                       \
      printf("%zu:", line - line + 1);                                         \
    if (args.opts.byteoff)                                                     \
      printf("%zu:", line - line + lineoff);                                   \
    if (args.opts.count)                                                       \
      printf("%zu\n", count);                                                  \
    else if (args.opts.list || args.opts.nlist)                                \
      printf("%s%c", *file, args.opts.nulname ? '\0' : '\n');                  \
  } while (0)

  int exit_status = EXIT_NOMATCH;
  int ieol = args.opts.oneline ? EOF : args.opts.nulline ? '\0' : '\n';

  for (char **file = args.files; *file; file++) {
    if (args.opts.quiet && exit_status == EXIT_MATCH)
      break; // at least one file has matches, no need to process more files

    FILE *fp = NULL;
    if (strcmp(*file, "-") == 0)
      fp = stdin;

#ifdef __unix__
    // on UNIX, when reading from a regular file and not stdin, memory-map the
    // file to improve performance
    else if (1) {
      int fd = open(*file, O_RDONLY);
      if (fd == -1)
        goto perror_continue;
      off_t size = lseek(fd, 0, SEEK_END);
      if (size == -1 ? close(fd), 1 : 0)
        goto perror_continue;
      uint8_t *data =
          size ? mmap(0, size, PROT_READ, MAP_PRIVATE, fd, 0) : &(uint8_t){0};
      if (data == MAP_FAILED ? close(fd), 1 : 0)
        goto perror_continue;

      size_t lineno = 0, count = 0, lineoff = 0;
      uint8_t *line = data, *p = data;

      for (; p < data + size; line = ++p) {
        struct dstate *dstate = dfa;
        while (!dstate->terminating && p < data + size && *p != ieol)
          dstate = dstate->transitions[*p++];
        if (p < data + size && *p != ieol)
          (p = memchr(p, ieol, data + size - p)) || (p = data + size);
        size_t len = p - line;

        if (!args.opts.oneline && p == data + size && len == 0)
          break; // ignore partial line if it's empty
        if (dstate->accepting && ++count) {
          OUTPUT_LINE;
          if (args.opts.quiet || exit_status != EXIT_ERROR)
            exit_status = EXIT_MATCH; // without '-q', EXIT_ERROR takes priority
          if (args.opts.list && !args.opts.count || args.opts.quiet)
            break; // file has at least one match, no need to process more lines
        }
        lineno++, lineoff += len + 1;
      }

      if (size && munmap(data, size) == -1 ? close(fd), 1 : 0)
        goto perror_continue;
      if (close(fd) == -1)
        goto perror_continue;

      OUTPUT_FILE;

      continue;
    }
#endif

    else if ((fp = fopen(*file, "r")) == NULL)
      goto perror_continue;

    size_t lineno = 0, count = 0, lineoff = 0;
    size_t len = 0, cap = 256;
    uint8_t *line = malloc(cap);

    for (; !feof(fp); len = 0) {
      struct dstate *dstate = dfa;
      for (int c; c = fgetc(fp), c != EOF && c != ieol; line[len++] = c) {
        len == cap ? line = realloc(line, cap *= 2) : 0;
        dstate = dstate->transitions[c];
      }
      if (ferror(fp) ? free(line), fclose(fp), 1 : 0)
        goto perror_continue;

      if (!args.opts.oneline && feof(fp) && len == 0)
        break; // ignore partial line if it's empty
      if (dstate->accepting && ++count) {
        OUTPUT_LINE;
        if (args.opts.quiet || exit_status != EXIT_ERROR)
          exit_status = EXIT_MATCH; // without '-q', EXIT_ERROR takes priority
        if (args.opts.list && !args.opts.count || args.opts.quiet)
          break; // file has at least one match, no need to process more lines
      }
      lineno++, lineoff += len + 1;
    }

    free(line);

    if (fp == stdin)
      clearerr(fp); // clear EOF in case file '-' is supplied more than once
    else if (fclose(fp) != 0)
      goto perror_continue;

    OUTPUT_FILE;

    continue;

  perror_continue:
    perror(*file);
    if (!args.opts.quiet || exit_status != EXIT_MATCH)
      exit_status = EXIT_ERROR; // with '-q', EXIT_MATCH takes priority
  }

  dfa_free(dfa);
  dfa_free(rev_dfa), dfa_free(fwd_dfa);

  return exit_status;
}
